engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java

```diff
@@ -0,0 +1,276 @@
+        Name providingModule = moduleManager.getEnvironment().getModuleProviding(this.getClass());
+
+        this.nodeUri = new SimpleUri(providingModule.toString() + ":" + nodeId);
+        //TODO Check for empty list of either in or out
+    }
+
+    /**
+     * Each node must implement this method to be called when the node is fully connected.
+     * This method should call other private node methods to use dependencies.
+     */
+    public abstract void setDependencies(Context context);
+
+    private void addInputConnection(DependencyConnection input) {
+        this.inputConnections.putIfAbsent(input.getName(), input);
+    }
+
+    private void addOutputConnection(DependencyConnection output) {
+        this.outputConnections.putIfAbsent(output.getName(), output);
+    }
+
+    /**
+     * TODO String to SimpleUri or make ConnectionUri and change Strings for names to ConnectionUris
+     */
+
+    protected FBO getInputFboData(int number) {
+        return ((FboConnection) this.inputConnections.get((new StringBuilder("FBO").append(number).toString()))).getFboData();
+    }
+
+    protected void addInputFboConnection(int id, FboConnection from) {
+        DependencyConnection fboConnection = new FboConnection(FboConnection.getFboName(id), DependencyConnection.Type.INPUT, from.getFboData());
+        addInputConnection(fboConnection);
+    }
+
+    protected void addOutputFboConnection(int id, FBO fboData) {
+        DependencyConnection fboConnection = new FboConnection(FboConnection.getFboName(id), DependencyConnection.Type.OUTPUT, fboData);
+        addOutputConnection(fboConnection);
+    }
+
+    public void connectFbo(int inputFboId, FboConnection from) {
+
+        addInputFboConnection(inputFboId, from);
+    }
+
+    public FboConnection getOutputFboConnection(int outputFboId) {
+        return (FboConnection) getOutputConnection(FboConnection.getFboName(outputFboId));
+    }
+
+    /*public FBO getOutputFboData(int id)
+        return ((FboConnection) this.outputConnections.get(DependencyConnection.getFboName(id))).getFboData();
+    }*/
+
+    public void removeInputConnection(String name) {
+        inputConnections.remove(name);
+    }
+
+    public void removeOutputConnection(String name) {
+        outputConnections.remove(name);
+    }
+
+    public int getNumberOfInputConnections() {
+        return this.inputConnections.size();
+    }
+
+    public int getNumberOfOutputConnections() {
+        return this.outputConnections.size();
+    }
+
+    private DependencyConnection getInputConnection(String name) {
+        return inputConnections.get(name);
+    }
+
+    private DependencyConnection getOutputConnection(String name) {
+        DependencyConnection connection = outputConnections.get(name);
+        if (connection == null) {
+            String errorMessage = String.format("Getting output connection named %s returned null. No such output connection" +
+                    "in %s", name, this.toString());
```

Comment:
Split this line on the period, and start the next one from "No such" to make the lines more self-contained and readable.